src/Features/Core/Portable/EmbeddedLanguages/StackFrame/StackFrameParser.cs

```diff
@@ -0,0 +1,455 @@
+        }
+
+        /// <summary>
+        /// Parses the array rank specifiers for an identifier. 
+        /// Ex: string[,][]
+        ///           ^----^ both are array rank specifiers
+        ///                  0: "[,]
+        ///                  1: "[]"
+        /// </summary>
+        private ParseResult<ImmutableArray<StackFrameArrayRankSpecifier>> ParseArrayRankSpecifiers()
+        {
+            using var _ = ArrayBuilder<StackFrameArrayRankSpecifier>.GetInstance(out var builder);
+            using var _1 = ArrayBuilder<StackFrameToken>.GetInstance(out var commaBuilder);
+
+            while (true)
+            {
+                if (!_lexer.ScanCurrentCharAsTokenIfMatch(StackFrameKind.OpenBracketToken, scanTrailingWhitespace: true, out var openBracket))
+                {
+                    return new(builder.ToImmutable());
+                }
+
+                while (_lexer.ScanCurrentCharAsTokenIfMatch(StackFrameKind.CommaToken, scanTrailingWhitespace: true, out var commaToken))
+                {
+                    commaBuilder.Add(commaToken);
+                }
+
+                if (!_lexer.ScanCurrentCharAsTokenIfMatch(StackFrameKind.CloseBracketToken, scanTrailingWhitespace: true, out var closeBracket))
+                {
+                    return ParseResult<ImmutableArray<StackFrameArrayRankSpecifier>>.Abort;
+                }
+
+                builder.Add(new StackFrameArrayRankSpecifier(openBracket, closeBracket, commaBuilder.ToImmutableAndClear()));
+            }
+
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        /// <summary>
+        /// Parses text for a valid file path using valid file characters. It's very possible this includes a path that doesn't exist but
+        /// forms a valid path identifier. 
+        /// 
+        /// Can return if only a path is available but not line numbers, but throws if the value after the path is a colon as the expectation
+        /// is that line number should follow.
+        /// </summary>
+        private ParseResult<StackFrameFileInformationNode> TryParseFileInformation()
+        {
+            var path = _lexer.TryScanPath();
+            if (!path.HasValue)
+            {
+                return ParseResult<StackFrameFileInformationNode>.Empty;
+            }
+
+            if (path.Value.Kind != StackFrameKind.PathToken)
+            {
+                return ParseResult<StackFrameFileInformationNode>.Abort;
+            }
+
+            if (!_lexer.ScanCurrentCharAsTokenIfMatch(StackFrameKind.ColonToken, out var colonToken))
+            {
+                return new(new(path.Value, null, null));
+            }
+
+            var lineNumber = _lexer.TryScanLineNumber();
+
+            // TryScanLineNumber can return a token that isn't a number, in which case we want 
+            // to bail in error and consider this malformed.
+            if (!lineNumber.HasValue || lineNumber.Value.Kind != StackFrameKind.NumberToken)
```

Comment:
> TryScanLineNumber can return a token that isn't a number

this seems highly suspect.  when/why is that a good hting?

Reply:
This is from returning `StackFrameToken.None`. With adding `StackFrameToken.InvalidNumberToken` I think this will make a lot more sense to explicitly check for that

Reply:
> This is from returning StackFrameToken.None

Hrmm... so i would expect you'd get null back in that case.  maybe i'm missing something.  What's teh case where you get a value back *and* it's the None result?

Reply:
There isn't one. It should be `Null | NumberToken | InvalidNumberToken`